File: #20_valid_parentheses/isValid.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const map = {
    "(": ")",
    "[": "]",
    "{": "}",
  };
  const stack = [];

  for (let c of s) {
    if (map[c]) {
      stack.push(c);
    } else {
      const char = stack.pop();
      if (map[char] !== c) {
        console.log(false);
        return false;
      }
    }
  }

  console.log(stack.length === 0);
  return stack.length === 0;
};

isValid("()"); // true
isValid("()[]{}"); // true
isValid("(]"); // false

File: stack/#20_valid_parentheses/isValid.go
package main

import "fmt"

func main() {
	isValid("()")     // true
	isValid("()[]{}") // true
	isValid("(]")     // false
	isValid("]")      // false
}

func isValid(s string) bool {
	stack := []rune{}
	brackets := map[rune]rune{
		'(': ')',
		'[': ']',
		'{': '}',
	}

	for _, c := range s {
		if v, ok := brackets[c]; ok {
			stack = append(stack, v)
		} else if len(stack) == 0 || stack[len(stack)-1] != c {
			fmt.Println(false)
			return false
		} else {
			stack = stack[:len(stack)-1]
		}
	}

	fmt.Println(len(stack) == 0)
	return len(stack) == 0
}
